src/sleep/sleep.c

```diff
@@ -44,7 +50,101 @@
+
+        if (!is_efi_boot())
+                return 0;
+
+        log_level = required ? LOG_ERR : LOG_DEBUG;
+        log_level_ignore = required ? LOG_WARNING : LOG_DEBUG;
+
+        r = sd_device_new_from_devnum(&device, 'b', hibernate_location->devno);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to create sd-device object for '%s': %m",
+                                      hibernate_location->swap->device);
+
+        r = sd_device_get_property_value(device, "ID_FS_UUID", &uuid_str);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to get filesystem UUID for device '%s': %m",
+                                      hibernate_location->swap->device);
+
+        r = sd_id128_from_string(uuid_str, &uuid);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to parse ID_FS_UUID '%s' for device '%s': %m",
+                                      uuid_str, hibernate_location->swap->device);
+
+        r = json_variant_new_uuid(&uuid_v, uuid);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to create JSON field for filesystem UUID: %m");
+
+        r = json_variant_set_field(&v, "uuid", uuid_v);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to add uuid to JSON object: %m");
+
+        r = json_variant_set_field_unsigned(&v, "offset", hibernate_location->offset);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to add offset to JSON object: %m");
+
+        r = parse_os_release(NULL,
+                             "ID", &id,
+                             "IMAGE_ID", &image_id,
+                             "VERSION_ID", &version_id,
+                             "IMAGE_VERSION", &image_version);
+        if (r < 0)
+                log_full_errno(log_level_ignore, r, "Failed to parse os-release, ignoring: %m");
+
+        if (id) {
+                r = json_variant_set_field_string(&v, "id", id);
+                if (r < 0)
+                        log_full_errno(log_level_ignore, r, "Failed to add id to JSON object, ignoring: %m");
+        }
+
+        if (image_id) {
+                r = json_variant_set_field_string(&v, "imageId", image_id);
+                if (r < 0)
+                        log_full_errno(log_level_ignore, r, "Failed to add imageId to JSON object, ignoring: %m");
+        }
+
+        if (version_id) {
+                r = json_variant_set_field_string(&v, "versionId", version_id);
+                if (r < 0)
+                        log_full_errno(log_level_ignore, r, "Failed to add versionId to JSON object, ignoring: %m");
+        }
+
+        if (image_version) {
+                r = json_variant_set_field_string(&v, "imageVersion", image_version);
+                if (r < 0)
+                        log_full_errno(log_level_ignore, r, "Failed to add imageVersion to JSON object, ignoring: %m");
+        }
```

Comment:
thinking about it we should probably include `uname -r` info here too and match it, i.e. kernel version.

i.e. uname()'s `release` field from struct utsname

Reply:
Well, actually you can resume to a different kernel (I've tried that several times). The control is given back to the kernel from memory image after resume: https://elixir.bootlin.com/linux/latest/source/kernel/power/hibernate.c#L545

But one thing I'm not sure about is that whether this has any sorts of guarantee for backwards compatibility...

Reply:
lets better be safe than sorry, and include the info, it's simple after all.